src/main/java/nl/b3p/tailormap/api/security/OIDCRepository.java

```diff
@@ -0,0 +1,95 @@
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+
+public class OIDCRepository implements ClientRegistrationRepository, Iterable<ClientRegistration> {
+  private static final Logger logger =
+      LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private final OIDCConfigurationRepository oidcConfigurationRepository;
+
+  private final Map<String, ClientRegistration> registrations;
+
+  public OIDCRepository(@Autowired OIDCConfigurationRepository repository) {
+    oidcConfigurationRepository = repository;
+    registrations = new HashMap<>();
+
+    synchronize();
+  }
+
+  @Override
+  public ClientRegistration findByRegistrationId(String registrationId) {
+    return registrations.get(registrationId);
+  }
+
+  @Override
+  public Iterator<ClientRegistration> iterator() {
+    return registrations.values().iterator();
+  }
+
+  public void synchronize() {
+    Map<String, ClientRegistration> newMap = new HashMap<>();
+
+    final HttpClient httpClient =
+        HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build();
+    for (OIDCConfiguration configuration : oidcConfigurationRepository.findAll()) {
+      try {
+        HttpRequest.Builder requestBuilder =
+            HttpRequest.newBuilder()
+                .uri(new URI(configuration.getIssuerUrl() + "/.well-known/openid-configuration"));
+        HttpResponse<String> response =
+            httpClient.send(requestBuilder.build(), HttpResponse.BodyHandlers.ofString());
+
+        OIDCProviderMetadata metadata = OIDCProviderMetadata.parse(response.body());
+
+        String id = String.format("%d", configuration.getId());
+        newMap.put(
+            id,
+            ClientRegistration.withRegistrationId(id)
+                .clientId(configuration.getClientId())
+                .clientSecret(configuration.getClientSecret())
+                .clientName(configuration.getName())
+                .scope("openid")
+                .issuerUri(metadata.getIssuer().toString())
+                .clientAuthenticationMethod(
+                    ClientAuthenticationMethod
+                        .CLIENT_SECRET_BASIC) // TODO: fetch from OIDC metadata
+                .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
+                .authorizationUri(metadata.getAuthorizationEndpointURI().toASCIIString())
+                .tokenUri(metadata.getTokenEndpointURI().toASCIIString())
+                .userInfoUri(metadata.getUserInfoEndpointURI().toASCIIString())
+                .providerConfigurationMetadata(metadata.toJSONObject())
+                .jwkSetUri(metadata.getJWKSetURI().toASCIIString())
+                .userNameAttributeName(configuration.getUserNameAttribute())
+                .redirectUri("{baseUrl}/api/oauth2/callback")
+                .build());
+      } catch (Exception e) {
+        logger.error("Failed to create OIDC client registration", e);
```

Comment:
For now an admin must check out the Tailormap logs to find out problems, we don't register whether a configured configuration was succesfully activated after loading the metadata. Not a problem for now but we could save the result of loading the configuration in the persistant object and show it in the admin with a button like 'reload' or something (see other comment about `synchronize()` method)

Reply:
Status field was added